doc/rm/ipgen_tool.md

```diff
@@ -0,0 +1,230 @@
+In the pre-processing ("rendering"), which is performed by the ipgen tool, templates of source files, written in the Mako templating language, are converted into "real" source files.
+The templates can be customized through template parameters, which are available within the templates.
+
+Ipgen is a command-line tool and a library.
+Users wishing to instantiate an IP template or query it for template parameters will find the command-line application useful.
+For use in higher-level scripting, e.g. within [topgen](topgen_tool.md) using ipgen as Python library is recommended.
+
+
+## Anatomy of an IP template
+
+An IP template is a directory with a well-defined directory layout, which mostly mirrors the standard layout of IP blocks.
+
+An IP template directory has a well-defined structure:
+
+* The IP template name (`<templatename>`) equals the directory name.
+* The directory contains a file `data/<templatename>.tpldesc.hjson` containing all configuration information related to the template.
+* The directory also contains zero or more files ending in `.tpl`.
+  These files are Mako templates and rendered into an file in the same location without the `.tpl` file extension.
+
+### The template description file
+
+Each IP template comes with a description itself.
+This description is contained in the `data/<templatename>.tpldesc.hjson` file in the template directory.
+The file is written in Hjson.
+
+It contains a top-level dictionary, the keys of which are documented next.
+
+#### List of template parameters: `template_param_list`
+
+Keys within `template_param_list`:
+
+* `name` (string): Name of the template parameter.
+* `desc` (string): Human-readable description of the template parameter.
+* `type` (string): Data type of the parameter. Valid values: `int`, `str`
+* `default` (string|int): The default value of the parameter. The data type should match the `type` argument. As convenience, strings are converted into integers on demand (if possible).
+
+#### Example template description file
+
+An exemplary template description file with two parameters, `src` and `target` is shown below.
+
+```hjson
+// data/<templatename>.tpldesc.hjson
+{
+  "template_param_list": [
+    {
+      "name": "src",
+      "desc": "Number of Interrupt Source",
+      "type": "int",
+      "default": "32"
+    },
+    {
+      "name": "target",
+      "desc": "Number of Interrupt Targets",
+      "type": "int",
+      "default": "32"
+    },
+  ],
+}
+```
+
+### Source file templates (`.tpl` files)
+
+Templates are written in the [Mako templating language](https://www.makotemplates.org/).
+All template parameters are available in the rendering context.
+For example, a template parameter `src` can be used in the template as `{{ src }}`.
+
+Furthermore, the following functions are available:
+
+* `unique_vlnv(vlnv)`: Transform a FuseSoC core name, expressed as VLNV string, into a unique name.
```

Comment:
Would instance_vlnv be better than unique_vlnv? Some how unique makes me think it would generate something different on each run.

Reply:
I agree, updated the code and the docs.